core/src/main/java/hudson/util/RobustReflectionConverter.java

```diff
@@ -371,6 +442,12 @@ public Object doUnmarshal(final Object result, final HierarchicalStreamReader re
             OldDataMonitor.report((Saveable)result, (ArrayList<Throwable>)context.get("ReadError"));
             context.put("ReadError", null);
         }
+        if (existingObject) {
+            for (Map.Entry<FieldExpectation, Object> entry : expectedFields.entrySet()) {
+                reflectionProvider.writeField(result, entry.getKey().getName(), entry.getValue(),
```

Comment:
Actually, it looks to me like it does - the expected field value is always uninitialized (see above discussion), and it's not checking whether the field was already seen, so it's overwriting with null/0/false/etc.